metered/src/num_wrapper.rs

```diff
@@ -0,0 +1,112 @@
+/// results in the original value.
+///
+/// Should we avoid calling `NumWrapper` with `count = 1`, i.e is the optimizer able
+/// to get rid of the wrapping computations?  The Godbolt compiler explorer shows that
+/// starting `opt-level = 1`, the generated call is equivalent.  Program below for
+/// future reference:
+/// ```rust
+/// pub fn wrap(count: usize) -> u8 {
+///     let fills: u8 = (count / u8::MAX as usize) as u8;
+///     let rem: u8 = (count % u8::MAX as usize) as u8;
+///     let v = rem.wrapping_sub(fills);
+///
+///     v
+/// }
+///
+/// pub fn naive(c: u8) -> u8 {
+///     c.wrapping_add(wrap(1))
+/// }
+///
+/// pub fn manual(c: u8) -> u8  {
+///     c.wrapping_add(1)
+/// }
+///
+///
+/// pub fn main(){
+///     let m = manual(42);
+///     let n = naive(42);
+///     println!("{n} {m}");
+/// }
+/// ```
+pub(crate) struct NumWrapper<T>(PhantomData<T>);
+
+macro_rules! impl_wrapping_metric_for {
+    ($int:path) => {
+        impl NumWrapper<$int> {
+            /// Booom
+            pub(crate) fn wrap(count: usize) -> $int {
+                let fills: $int = (count / <$int>::MAX as usize) as $int;
+                let rem: $int = (count % <$int>::MAX as usize) as $int;
+                let v = (rem.wrapping_sub(fills));
+
+                v
+            }
+        }
+    };
+}
+
+impl_wrapping_metric_for!(u8);
+impl_wrapping_metric_for!(u16);
+impl_wrapping_metric_for!(u32);
+impl_wrapping_metric_for!(u64);
+impl_wrapping_metric_for!(u128);
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use proptest::prelude::*;
+
+    // test with u8 for more wrapping - same model/properties applies to other ints.
+    fn incr_by(cc: u8, count: usize) -> u8 {
+        let v = NumWrapper::<u8>::wrap(count);
+        cc.wrapping_add(v)
+    }
+    fn incr_by_naive(mut cc: u8, count: usize) -> u8 {
+        for _ in 0..count {
+            cc = cc.wrapping_add(1);
+        }
+        cc
+    }
+
+    fn decr_by(cc: u8, count: usize) -> u8 {
+        let v = NumWrapper::<u8>::wrap(count);
+        cc.wrapping_sub(v)
+    }
+    fn decr_by_naive(mut cc: u8, count: usize) -> u8 {
+        for _ in 0..count {
+            cc = cc.wrapping_sub(1);
+        }
+        cc
+    }
+
+    proptest! {
+        #[test]
+        fn test_wrapping(x: u8, y: u16) {
+            // reduce strategy space, usize takes too long
+            let y = y as usize;
+            // Tests if calling incr() Y times returns the same value
+            // as the optimized version
+            assert_eq!(incr_by_naive(x, y), incr_by(x, y));
+            // Tests if calling decr() Y times returns the same value
+            // as the optimized version
+            assert_eq!(decr_by_naive(x, y), decr_by(x, y));
+            // Tests if calling decr() Y times on incr() Y times returns
+            // the original value
+            assert_eq!(x, decr_by(incr_by(x, y), y));
+        }
+    }
```

Comment:
That's super-nice. That's the first time I read about `proptest`, and I saw that it is possible to specify a range of values for the inputs, so I'm wondering if we could get rid of the casting by just limiting the range for the usize to that of u16 or something alike.

https://altsysrq.github.io/proptest-book/proptest/tutorial/macro-proptest.html

Reply:
Thanks, I wanted to do that but didn't have time to dig in the docs to figure out how to do that 'properly'

Reply:
In case you don't have much time right now, you can just merge this suggestion. I already tested it in my laptop:

```suggestion
    proptest! {
        #[test]
        fn test_wrapping(x: u8, y in 0..4096usize) {
            // Tests if calling incr() Y times returns the same value
            // as the optimized version
            assert_eq!(incr_by_naive(x, y), incr_by(x, y));
            // Tests if calling decr() Y times returns the same value
            // as the optimized version
            assert_eq!(decr_by_naive(x, y), decr_by(x, y));
            // Tests if calling decr() Y times on incr() Y times returns
            // the original value
            assert_eq!(x, decr_by(incr_by(x, y), y));
        }
    }
```